Refactor code to use Argparse instead of GetOpt for improved readability + flexibility

Iridium-IO opened the pull request:
**Argparse** 
 - Commandline parameters now use `argparse()` instead of `getopt()`. This reduces the need to provide separate help/default/variable assignments and makes it easier to manage new parameters. This change is feature-equal to the old `getopt` version, the only real difference being that command-line arguments now always override config-file arguments
     -  See: https://github.com/arpruss/gcodeplot/pull/43#issuecomment-2819672489
 - Custom argparse definitions and functions have been allocated to `gcodeplotutils/argparse_c`. This mainly has custom extensions to handle features requires for parity with the old system (comments in config files, `--no` prefixes, etc) 

**Inkscape**
 - GCode can now be exported using `File > Export` in Inkscape. This allows you to keep a degree of separation between the SVG and exported GCode (plus it's fewer clicks when rapidly testing, as you can keep the Export window open in a tab. Saving normally still works as expected. 
 - The layout for the GCodePlot dialog has been improved with better grouping of related variables
 - `Extract only one color from drawing` now uses the new `Color` selector in Inkscape instead of passing RGB values as a string

![1](https://github.com/arpruss/gcodeplot/assets/1491536/36d6797f-ecff-4a1c-89bb-eabffa19564b)
![3](https://github.com/arpruss/gcodeplot/assets/1491536/1db97ef5-24d0-4c68-8dad-18955addb3a3)
![2](https://github.com/arpruss/gcodeplot/assets/1491536/164488ac-2e0f-413d-9f42-366f386488cf)

**General**
- Simplified `__main__` to reduce nesting and make code easier to read
- Added comments to `__main__` describing different sections


**Note**
This PR also includes the changes from #42 adding Moonraker support. Sorry, I didn't quite know how to separate the changes without deleting all the Moonraker stuff, creating this PR, then re-adding the Moonraker stuff. I don't want to clog up your repo with changes and un-changes, but if you're more happy with this PR (as it's more testable since you don't have Moonraker), I can do the following steps:
  - Remove Moonraker support from within this branch (IridiumIO/GcodePlot/argparse) which will automatically update this PR
  - Merge this PR into your base repository if accepted
  - Pull these changes into the Moonraker PR which can then sit open until someone else tests it. 
 
Thoughts? I've tested the generated GCode using my own tests (complex overlapped shapes with variable shading/cutting) as well as running the included `mpcnc` and `cutter` config files to check that the output gcode is identical. There is some variation in _both_ the new and old files due to non-deterministic choosing of which shapes to draw first (e.g. one time it will draw `star > circle > square` then the next time you run the script it will draw `circle > star > square`

Iridium-IO committed:
Add Moonraker API Config

Added Moonraker config, and separated connection settings (Moonraker, Serial) to its own tab

Iridium-IO committed:
Added Moonraker API support

GCode can be uploaded via Moonraker (e.g. to Klipper). Currently does not have authentication support.

Iridium-IO committed:
- Refactor to use argparse instead of getopt to minimise code duplication
    - Removes manual "help" and "default" console commands and uses argparse generators instead.
    - config files are now passed directly into the command line by using a '$' prefix, e.g. gcodeplot.py $config.txt
    - most '--no' prefixes are now reduntant. Some are still retained but I don't really know why they were needed in the first place. Renamed some arguments to be clearer in this regard.
- Add type hinting for some parameters

Issues:
 - for some reason the 'cut' mode doesn't generate the exact same ordering of overlapping cuts anymore. All cuts still execute correctly, just the order of shapes being cut is different. To be investigated.

Iridium-IO committed:
GCode can now be exported using the File > Export menu (to avoid constantly having to use "Save As"
Updated layout of Inkscape config menu.
  - Added separators and groups to collect similar settings together visually
  - Now uses type="color" for the extract-color option, allowing visual picking of a color instead of typing in a string.

Iridium-IO committed:
- Removed uneccesary argument creation
- Separated svgTree parsing

Iridium-IO committed:
Update rgbFromColor() to accept the new handling of colors from Inkscape as uint32 numbers.

Iridium-IO committed:
Undid the previous change to use "-1" instead of "none" for null direction angle values. Now accepts "none" at the commandline again
 - possible that this was causing the variation in the order that shapes were being cut when using 'cut' mode.

Iridium-IO committed:
Split penData creation into separate function

Iridium-IO committed:
Add missing return in generate_pen_data
Clean up imports

Iridium-IO committed:
Remove unecessary(?) variable assignment

Iridium-IO committed:
Fixed shading and color extraction not working in 'draw' mode.

Iridium-IO committed:
- Breakout argparse functions and classes into separate file
- Cleanup unecessary arguments and enable input parity with old "--no" prefixes
- Create enums.py so that argparser_c.py can use it too.
- Tidy up __main__ section

Iridium-IO committed:
- Breakout argparse functions and classes into separate file
- Cleanup unecessary arguments and enable input parity with old "--no" prefixes
- Create enums.py so that argparser_c.py can use it too.
- Tidy up __main__ section

Iridium-IO committed:
Merge branch 'argparse' of https://github.com/IridiumIO/GCodePlot into argparse

Iridium-IO committed:
Fix argument handling of args.sort and args.optimization_time
- Optimization time was broken due to using false-y boolean check rather than properly checking for True/False values

arpruss commented:
I would love to merge everything other than the argparse, because the change in config file usage could break somebody's shell script.

Iridium-IO commented:
I'll see if I can figure out how to keep it consistent so old scripts work too :) 

I probably should've use optparse instead of argparse, but I didn't even know that existed at the time

Iridium-IO committed:
Adjust argument parsing to support configuration files passed in using the old getopt() style

- If both command-line parameters and a config file are used, the command-line parameters will overwrite ones from the config file

Iridium-IO committed:
Merge branch 'master' into argparse

Iridium-IO commented:
@arpruss okay the old way of using config files should work again now, I've tested using the included mpcnc files and they work. 

I did make one small change, though: command-line arguments now **always override** values in the config file, regardless of the order they're provided. Previously, the behavior depended on the order the `--config-file` argument appeared.

Given a config file:
```cfg
pen-down-speed=5
```

Old outcome:
```ps1
gcodeplot.py --pen-down-speed 10 --config-file cfg
>> pen-down-speed = 5

gcodeplot.py --config-file cfg  --pen-down-speed 10
>> pen-down-speed = 10
```

New outcome:
```bat
gcodeplot.py --pen-down-speed 10 --config-file cfg
>> pen-down-speed = 10

gcodeplot.py --config-file cfg  --pen-down-speed 10
>> pen-down-speed = 10
```

This probably won't matter for most users (since mixing config files and CLI args probably isn't common), but I think it's more intuitive: if someone explicitly passes an argument on the command line, it should take precedence over the config file.

polhomarkho commented:
Hello! Any updates on this PR? As it's a lot of changes and improvements, I would love to use it as a base to work on features I would like to try and implement. If you need some help testing, I can help 🙂

arpruss commented:
I am sorry. I just don't love the argument parsing semantics getting changed. I think arguments should be parsed left-to-right and later arguments should always override earlier ones.

Iridium-IO commented:
> I am sorry. I just don't love the argument parsing semantics getting changed. I think arguments should be parsed left-to-right and later arguments should always override earlier ones.

They do work that way, with the exception of passing in a config file alongside command-line arguments. I think passing an argument to the command line means that one should take precedence over any hidden in a file, right? 

Otherwise it would be confusing if you're intending to override an argument from the file, but declaring the override argument before the config file parameter causes it to fail.
